File: backend/app/search/normalize.py
"""The one normalization path.

Every comparison this product makes between "what a person typed" and "what is
on the books" goes through this module. There is no ``.lower()`` and no
whitespace rule anywhere else in the search code, because two normalizations
that drift apart produce a search that finds a customer on one screen and not on
another — and a resolver that identifies the wrong person is worse than one that
finds nobody.

**Normalization never rewrites stored data for display.** ``normalized_name`` and
``customer_alias.normalized`` sit *beside* the real name and the real alias; the
person's name is shown exactly as it was entered, always. What is normalized is
only ever the comparison key.

What it does, in order:

1. NFKD decompose and drop combining marks, so ``Ayesha`` matches ``Áyesha`` and
   an Urdu word matches the same word written with harakat;
2. NFKC recompose, so compatibility forms (full-width digits, ligatures) fold to
   their ordinary equivalents;
3. casefold — stronger than ``lower()`` and defined for non-Latin scripts;
4. replace every non-alphanumeric character with a space, so ``Ahmed-bhai``,
   ``Ahmed_bhai`` and ``Ahmed  bhai`` all become one comparison key;
5. collapse runs of whitespace and strip.

``str.isalnum`` is true for Urdu, Arabic and Devanagari letters, so a name stored
in its own script normalizes to itself and matches when typed in that script.

**What it deliberately does not do.** There is no transliteration between scripts
and no Roman-Urdu spelling model: ``احمد`` does not match ``Ahmed`` unless one of
them is stored as an alias of the customer. That is the whole point of aliases —
a person records the name they actually use, and the system stops guessing. A
transliteration engine would need language tooling, would be wrong for names in
ways nobody could predict, and would make identification less trustworthy rather
than more.
"""

from __future__ import annotations

import unicodedata

__all__ = [
    "MAX_QUERY_LENGTH",
    "PHONE_SUFFIX_DIGITS",
    "PHONE_SUFFIX_MIN_DIGITS",
    "looks_like_phone",
    "normalize_phone",
    "normalize_text",
    "normalize_tokens",
    "phone_suffix",
]

#: Longer than any name on the books, and short enough that no query can be used
#: to make the database do unbounded work. Queries are truncated, never rejected:
#: a paste accident should find nothing useful, not raise.
MAX_QUERY_LENGTH = 120

#: The shortest digit string that may be matched as a phone *suffix*. Below this
#: a "phone" match would be a coincidence — three digits match half the book.
PHONE_SUFFIX_MIN_DIGITS = 7

#: How many trailing digits a suffix match compares.
#:
#: Numbers are stored E.164 (``+923001234567``) and people type them however
#: they hold them — often in the national form (``0300-1234567``), where a
#: leading ``0`` stands in for the country code. Comparing the *trailing* nine
#: digits makes both forms find the same person without this module knowing a
#: single thing about any country's dialling plan. Nine digits is specific enough
#: that a collision would have to be arranged on purpose.
PHONE_SUFFIX_DIGITS = 9


def normalize_text(value: str | None) -> str:
    """The comparison key for a name, an alias, a code or an area."""
    if not value:
        return ""
    text = value[:MAX_QUERY_LENGTH]
    decomposed = unicodedata.normalize("NFKD", text)
    stripped = "".join(ch for ch in decomposed if not unicodedata.combining(ch))
    folded = unicodedata.normalize("NFKC", stripped).casefold()
    spaced = "".join(ch if ch.isalnum() else " " for ch in folded)
    return " ".join(spaced.split())


def normalize_tokens(value: str | None) -> tuple[str, ...]:
    """The normalized words of ``value``, in the order they were written.

    Order is preserved but never *required*: token matching compares sets, which
    is what makes "Ahmed bhai" and "bhai Ahmed" the same query.
    """
    normalized = normalize_text(value)
    return tuple(normalized.split()) if normalized else ()


def normalize_phone(value: str | None) -> str:
    """Every digit in ``value``, in order, with nothing else.

    ``+92 300 123-4567``, ``0092-3001234567`` and ``+923001234567`` all reduce to
    a digit string, so a person may type a number however they hold it. Non-ASCII
    digits (Eastern Arabic, Devanagari) fold to their ASCII value rather than
    being discarded, because a phone number typed in Urdu digits is still that
    phone number.
    """
    if not value:
        return ""
    digits: list[str] = []
    for ch in value[:MAX_QUERY_LENGTH]:
        if ch.isdigit():
            try:
                digits.append(str(unicodedata.digit(ch)))
            except (TypeError, ValueError):  # pragma: no cover - isdigit implies digit
                continue
    return "".join(digits)


def phone_suffix(value: str | None) -> str:
    """The trailing digits a suffix comparison uses. Empty when too short."""
    digits = normalize_phone(value)
    if len(digits) < PHONE_SUFFIX_MIN_DIGITS:
        return ""
    return digits[-PHONE_SUFFIX_DIGITS:]


def looks_like_phone(value: str | None) -> bool:
    """True when ``value`` is plausibly a phone number rather than a name.

    Deliberately strict: any letter at all means it is a name. ``+92 300`` is a
    phone fragment; ``Ahmed 3`` is not.
    """
    if not value:
        return False
    if any(ch.isalpha() for ch in value):
        return False
    return len(normalize_phone(value)) >= PHONE_SUFFIX_MIN_DIGITS

File: frontend/src/search/normalize.ts
/**
 * The client's single normalization path — and an honest account of its limits.
 *
 * This is a *mirror* of `app/search/normalize.py`, and it exists for exactly one
 * reason: offline, there is no server to ask, and the device still has to be
 * able to find "Ahmed bhai" in the customers it has already synchronised. It is
 * used by nothing else. Every online search and every identification goes to the
 * server, where the authoritative implementation lives.
 *
 * It performs the same five steps, in the same order:
 *
 *  1. NFKD decompose and drop combining marks, so `Ayesha` matches `Áyesha`;
 *  2. NFKC recompose, folding compatibility forms;
 *  3. lower-case;
 *  4. replace every non-letter, non-digit with a space, so `Ahmed-bhai`,
 *     `Ahmed_bhai` and `Ahmed  bhai` become one comparison key;
 *  5. collapse whitespace and trim.
 *
 * **Where it differs from the server, and why that is safe.** Python's
 * `str.casefold()` is stronger than JavaScript's `toLowerCase()` — German `ß`
 * folds to `ss` there and stays `ß` here — so a handful of exotic spellings can
 * normalize differently on the two sides. That divergence can only ever make
 * *offline* search miss a row the server would have found; it can never make it
 * find the wrong person, because the answer offline is always a list somebody
 * picks from. Online, this function is not consulted at all. If that trade ever
 * stops being acceptable the fix is a full case-folding table here, not a second
 * matching implementation.
 *
 * **Display text is never rewritten.** Everything below produces a comparison
 * key. The name on screen is always exactly what the owner typed.
 */

/** Longer than any name on the books; the same bound the server applies. */
export const MAX_QUERY_LENGTH = 120;

/** The shortest digit string that may be matched as a phone suffix. */
export const PHONE_SUFFIX_MIN_DIGITS = 7;

/** How many trailing digits a suffix match compares. */
export const PHONE_SUFFIX_DIGITS = 9;

const COMBINING = /\p{M}/gu;
const NOT_ALPHANUMERIC = /[^\p{L}\p{N}]+/gu;

/** The comparison key for a name, an alias, a code or an area. */
export function normalizeText(value: string | null | undefined): string {
  if (!value) return "";
  return value
    .slice(0, MAX_QUERY_LENGTH)
    .normalize("NFKD")
    .replace(COMBINING, "")
    .normalize("NFKC")
    .toLowerCase()
    .replace(NOT_ALPHANUMERIC, " ")
    .trim()
    .replace(/\s+/g, " ");
}

/**
 * The normalized words of `value`.
 *
 * Order is preserved but never required: token matching compares memberships,
 * which is what makes "Ahmed bhai" and "bhai Ahmed" the same query.
 */
export function normalizeTokens(value: string | null | undefined): string[] {
  const normalized = normalizeText(value);
  return normalized ? normalized.split(" ") : [];
}

/**
 * Every digit in `value`, in order, with nothing else.
 *
 * `+92 300 123-4567` and `0092-3001234567` both reduce to a digit string, so a
 * number may be typed however it is held.
 */
export function normalizePhone(value: string | null | undefined): string {
  if (!value) return "";
  let digits = "";
  for (const ch of value.slice(0, MAX_QUERY_LENGTH)) {
    // Unicode-aware: an Eastern Arabic or Devanagari digit is still that digit.
    const decimal = ch.normalize("NFKD").replace(COMBINING, "");
    if (decimal.length === 1 && decimal >= "0" && decimal <= "9") digits += decimal;
  }
  return digits;
}

/** The trailing digits a suffix comparison uses. Empty when too short. */
export function phoneSuffix(value: string | null | undefined): string {
  const digits = normalizePhone(value);
  if (digits.length < PHONE_SUFFIX_MIN_DIGITS) return "";
  return digits.slice(-PHONE_SUFFIX_DIGITS);
}

/**
 * True when `value` is plausibly a phone number rather than a name.
 *
 * Strict on purpose: any letter at all means it is a name.
 */
export function looksLikePhone(value: string | null | undefined): boolean {
  if (!value) return false;
  if (/\p{L}/u.test(value)) return false;
  return normalizePhone(value).length >= PHONE_SUFFIX_MIN_DIGITS;
}
